File: CS106A/BookExercisesChapter08 (Strings and Characters)/PigLatin.java
/*
 * File: PigLatin.java
 * Name: Oliver Pecha
 * Section Leader: Online Learning
 * -----------------
 * Book / Chapter 8 / Programming Exercise 17 & 18
 * -------------------
 * 17:  Rewrite the PigLatin program in this chapter so that it uses the algorithmic strategy of going through the line character by character 
 *    	rather than word by word. The pseudocode version of this strategy appears in the section on “Talking spaces and punctuation into account.” 
 * 18:  As written, the PigLatin program in Figure 8-6 behaves somewhat oddly  if you enter a string that includes words that begin with an 
 * 		uppercase letter. For example if you were to capitalize the first word in the sentence and the name of the Pig latin language, you would 
 * 		see the following output:
 * -------------------
 * This program translates a line into Pig Latin.
 * Enter a line: This is Pig Latin.
 * isThay isway igPay atinLay.
 * -------------------
 * 		Rewrite the translateWord method so that any word that begins with a capital letter in the English line still begins with a capital 
 * 		letter in Pig Latin. Thus, after making your changes in the program, the output should look like this:
 * -------------------
 * This program translates a line into Pig Latin. 
 * Enter a line: This is Pig Latin.
 * Isthay isway Igpay Atinlay.
 * -------------------
 */

import acm.program.*;
import java.util.*;

public class PigLatin extends ConsoleProgram {

	public void run() {
		while (true) {
		println("\nThis program translates a line into Pig Latin.");
		
		String line = readLine("Enter a line: (This is Pig Latin.) ");
		println("\"" + translateLine(line) + "\"");
		}
	}

/* Translates a line to Pig Latin, word by word */
	private String translateLine(String line) {
		// Initialize a variable to hold the result and set it to the empty string.
		System.out.println("lenght is: " + line.length());
		String result = "";
		String extractedWord = "";
		int StartingChar = 0;
		for (int i = 0; i < line.length(); i++) {
			System.out.println("i is: " + i + "StartingChar is: " + StartingChar);
			// If the ith character in the line is some kind of separator
			if (isSeparator(line.charAt(i))) {
				//Append that character to the result variable.
				result = result + line.charAt(i);
				System.out.println("if isSeparator, result is: " + result);
				// If not the last char in the String, reset StartingChar for wordExtractor to trim correctly
				if (i + 1 < line.length()) {
				StartingChar = i + 1;
				} 
			}
			// Else If, the ith character is the end of a word
			else if (i + 1 < line.length()) {	
				if (isSeparator(line.charAt(i + 1))) {
					// Extract the word as a substring. 		
					extractedWord = wordExtractor(line, StartingChar, i + 1);
					// Translate the word to Pig Latin.	
					// Append the translated word to the result variable.
					result = result + translateWord(extractedWord);
				}
			}
			// Else If, the ith character is the end of the String Line entered
			else if (i + 1 == line.length()) {
				extractedWord = wordExtractor(line, StartingChar, i + 1);
				result = result + translateWord(extractedWord);
			}
		}
		// Return the value of the result variable to the caller.
		return result;
	}
	
	private String wordExtractor(String line, int StartingChar, int i) {
		String extractedWord = line.substring(StartingChar, i);
		System.out.println("extractedWord is: \"" + extractedWord + "\"");
		return extractedWord;
	}
/* Translates a line to Pig Latin, word by word */
	private String translateLineOld(String line) {
		String result = "";
		StringTokenizer tokenizer =
		  new StringTokenizer(line, DELIMITERS, true);
		while (tokenizer.hasMoreTokens()) {
			String token = tokenizer.nextToken();
			if (isWord(token)) {
				token = translateWord(token);
			}
			result += token;
		}
		return result;
	}
	
	private boolean isSeparator(char currentChar) {
		boolean result = false;
		switch (currentChar) {
		case ',': 
		case '.': 
		case '?':
		case ' ': result = true;
		}
		return result;
	}

/* Returns true if token is a "word" (all character are letters) */
	private boolean isWord(String token) {
		for (int i = 0; i < token.length(); i++) {
			char ch = token.charAt(i);
			if (!Character.isLetter(ch)) return false;
		}
		return true;
	}
/* Translates a word to Pig Latin and returns the translated word */
	private String translateWord(String word) {
		String result = "";
		int vp = findFirstVowel(word);
		if (vp == -1) {
			if (Character.isUpperCase(word.charAt(0))) {
				word = capitalizedWord(word);
			}
			result = word;
		} else if (vp == 0) {
			if (Character.isUpperCase(word.charAt(0))) {
				word = capitalizedWord(word);
			}
			result = word + "way";
		} else {
			String head = word.substring(0, vp);
			String tail = word.substring(vp);
			if (Character.isUpperCase(word.charAt(0))) {
				tail = capitalizedWord(tail);
				head = head.toLowerCase();
			}
			result = tail + head + "ay";
		}
		return result;
	}

/* Returns a word that has been Capitalized, regardless of the previous casing */
	private String capitalizedWord(String word) {
		word = word.toLowerCase();
		String capitalized = null;
		capitalized = "" + word.charAt(0);
		capitalized = capitalized.toUpperCase();
		capitalized = capitalized + word.substring(1);
		return capitalized;
	}
	
/* Translates a word to Pig Latin and returns the translated word */
	private String translateWordOld(String word) {
		int vp = findFirstVowel(word);
		if (vp == -1) {
			return word;
		} else if (vp == 0) {
			return word + "way";
		} else {
			String head = word.substring(0, vp);
			String tail = word.substring(vp);
			return tail + head + "ay";
		}
	}

/* Returns the index of the first vowel in the word (-1 if none) */
	private int findFirstVowel(String word) {
		for (int i = 0; i < word.length(); i++) {
			if (isEnglishVowel(word.charAt(i))) return i;
		}
		return -1;
	}

/* Returns true if the character is a vowel */
	private boolean isEnglishVowel(char ch) {
		switch (Character.toLowerCase(ch)) {
			case 'a': case 'e': case 'i': case 'o': case 'u':
				return true;
			default:
				return false;
		}
	}

/* Defines the characters that delimit a token */
	private static final String
	  DELIMITERS = "!@#$%^&*()_-+={[}]:;\"'<,>.?/~`|\\";

}

File: CS106B/Chapter03 (Strings) Book Exercises/15. PigLatin/PigLatin.cpp
/*
 * File: PigLatin.cpp
 * --------------
 * As written, the PigLatin program in Figure 3-2 behaves oddly if you enter a string
 * that includes words beginning with an uppercase letter. For example, if you were
 * to capitalize the first word in the sentence and the name of the Pig Latin language,
 * you would see the following output:
 *
 * Rewrite the wordToPigLatin function so that any word that begins with a capital letter
 * in the English line still begins with a capital letter in Pig Latin. Thus, after making
 * your changes in the program, the output should look like this:
 *
 */

#include "console.h"
#include "simpio.h"
using namespace std;


string PigLatin(string word);
string lineToPigLatin (string line);
string wordToPigLatin (string word);
int findFirstVowel(string word);
bool isVowel (char ch);
string capitalize(string str);

const string SENTINEL = "";

int main() {
    string sentence = " ";
    cout << "This program translates Englisg to PigLatin. \nIndicate the end of the input with a blank line.. " << endl;
    while (sentence != SENTINEL) {
        sentence = getLine("\nEnter English text:");
        cout << "PigLatin output: " << lineToPigLatin(sentence) << "." << endl;
    }
    return 0;
}

/*
 * Function: PigLatin
 * Usage: translation = PigLatin(word);
 * ------------------------------------
 * This function translates a word from English to Pig Latin using
 * the rules specified in Chapter 3.  The translated word is
 * returned as the value of the function.

string PigLatin(string word) {
    int vp = findFirstVowel(word);
    if (vp == -1) {
        return word;
    } else if (vp == 0) {
        return word + "way";
    } else {
        string head = word.substr(0, vp);
        string tail = word.substr(vp);
        return tail + head + "ay";
    }
}
 */






/*
* Function: lineToPigLatin
* Usage: string translation = lineToPigLatin (line);
* --------------------------------
* Translates each word in the line to Pig Latin, leaving all other
* characters unchanged. The variable start keeps track of the index
* position at which the current word begins. As a special case,
* the code sets start to -1 to indicate that the beginning of the
* current word has not yet been encountered.
*/
string lineToPigLatin (string line) {
    string result;
    int start = -1;
    for (int i = 0; i < line.length(); i++) {
        char ch = line[i];
        if (isalpha (ch)) {
            if (start == -1) start = i;
        } else {
            if (start >= 0) {
                result += wordToPigLatin(line.substr(start, i - start));
                start = -1;
            }
            result += ch;
            }
    }
    if (start >= 0) result += wordToPigLatin (line.substr(start));
    return result;
}

/*
* Function: wordToPigLatin
* Usage: string translation = wordToPigLatin (word);
* --------------------------------
* This function translates a word from English to Pig Latin using
* the rules specified in the text. The translated word is
* returned as the value of the function.
*/

string wordToPigLatin (string word) {
    int vp = findFirstVowel(word);
    string result;
    if (vp == -1) {
        result = word;
    } else if (vp == 0) {
        result = word + "way";
    } else {
        string head = word.substr(0, vp);
        string tail = word.substr(vp);
        result = tail + head + "ay";
    }
    return capitalize(result);
}


/*
 * Function: FindFirstVowel
 * Usage: k = FindFirstVowel(word);
 * --------------------------------
 * This function returns the index position of the first vowel
 * in word.  If word does not contain a vowel, FindFirstVowel
 * returns -1. The code for IsVowel appears in Chapter 1.
 */
int findFirstVowel(string word) {
    for (int i = 0; i < word.length(); i++) {
        if (isVowel(word[i])) return i;
    }
return -1; }

/*
 * Function: isVowel
 * Usage: if (= FindFirstVowel(word))isVowel(ch)) . . .
 * --------------------------------
 * Returns true if the character ch is a vowel.
 */

bool isVowel (char ch) {
        switch (ch) {
            case 'A': case 'E': case 'I': case 'O': case 'U':
        case 'a': case 'e': case 'i': case 'o': case 'u':
            return true;
        default:
            return false;
        }
}


string capitalize(string str) {
    int index = 0;
    for (int i = 0; i < str.length(); i++){
        if (index == 0) {
           str[i] = toupper(str[i]);
           index++;
        }
        else {
            str[i] = tolower(str[i]);
            index++;
        }
        if (isspace(str[i+1])) {
            index = -1;
        }
    }
    return str;
}
